File: basic_dsa_one_hundred_fifty.java
/**
 * Basic DSA Problems - Base 7
 * Convert integer to base 7 string
 */
public class basic_dsa_one_hundred_fifty {

    // Problem: Given an integer num, return its string representation in base 7.
    // Example: 100 -> "202", -7 -> "-10"

    /**
     * Convert by repeatedly dividing by 7, collecting remainders
     * Time O(log n), Space O(log n) for output
     */
    public static String convertToBase7(int num) {
        if (num == 0) return "0";
        boolean neg = num < 0;
        num = Math.abs(num);
        StringBuilder sb = new StringBuilder();
        while (num > 0) {
            sb.append(num % 7);
            num /= 7;
        }
        if (neg) sb.append('-');
        return sb.reverse().toString();
    }

    public static void main(String[] args) {
        System.out.println("=".repeat(60));
        System.out.println("Problem: Base 7");
        System.out.println("=".repeat(60));

        System.out.println("100 -> " + convertToBase7(100));
        System.out.println("-7 -> " + convertToBase7(-7));
        System.out.println("0 -> " + convertToBase7(0));
        System.out.println("7 -> " + convertToBase7(7));
    }
}

File: basic_dsa_one_hundred_fifty.py
"""
Basic DSA - Maximum Subarray (Kadane's Algorithm)
Given an integer array nums, find the contiguous subarray with the largest sum and
return its sum.
"""


def max_subarray(nums):
    """
    Kadane's algorithm: best subarray ending here vs start new.
    Time Complexity: O(n)
    Space Complexity: O(1)
    """
    best = nums[0]
    cur = nums[0]
    for x in nums[1:]:
        cur = max(x, cur + x)
        best = max(best, cur)
    return best


if __name__ == "__main__":
    print("=" * 60)
    print("Problem: Maximum Subarray")
    print("=" * 60)

    print(max_subarray([-2, 1, -3, 4, -1, 2, 1, -5, 4]))  # 6
    print(max_subarray([1]))  # 1
    print(max_subarray([5, 4, -1, 7, 8]))  # 23
